workflow/artifacts/git/git.go

```diff
@@ -112,111 +77,87 @@ func (g *ArtifactDriver) Save(string, *wfv1.Artifact) error {
 func (g *ArtifactDriver) Load(inputArtifact *wfv1.Artifact, path string) error {
-	sshUser := GetUser(inputArtifact.Git.Repo)
-	closer, auth, env, err := g.auth(sshUser)
+	a := inputArtifact.Git
+	sshUser := GetUser(a.Repo)
+	closer, auth, err := g.auth(sshUser)
 	if err != nil {
 		return err
 	}
 	defer closer()
-
-	var recurseSubmodules = git.DefaultSubmoduleRecursionDepth
-	if inputArtifact.Git.DisableSubmodules {
-		log.Info("Recursive cloning of submodules is disabled")
-		recurseSubmodules = git.NoRecurseSubmodules
-	}
-	repo, err := git.PlainClone(path, false, &git.CloneOptions{
-		URL:               inputArtifact.Git.Repo,
-		RecurseSubmodules: recurseSubmodules,
-		Auth:              auth,
-		Depth:             inputArtifact.Git.GetDepth(),
-	})
+	depth := a.GetDepth()
+	r, err := git.PlainClone(path, false, &git.CloneOptions{URL: a.Repo, Auth: auth, Depth: depth})
 	switch err {
 	case transport.ErrEmptyRemoteRepository:
-		log.Info("Cloned an empty repository ")
+		log.Info("Cloned an empty repository")
 		r, err := git.PlainInit(path, false)
 		if err != nil {
-			return err
+			return fmt.Errorf("failed to plain init: %w", err)
 		}
-		if _, err := r.CreateRemote(&config.RemoteConfig{Name: git.DefaultRemoteName, URLs: []string{inputArtifact.Git.Repo}}); err != nil {
-			return err
+		if _, err := r.CreateRemote(&config.RemoteConfig{Name: git.DefaultRemoteName, URLs: []string{a.Repo}}); err != nil {
+			return fmt.Errorf("failed to create remote %q: %w", a.Repo, err)
 		}
-		branchName := inputArtifact.Git.Revision
+		branchName := a.Revision
 		if branchName == "" {
 			branchName = "master"
 		}
 		if err = r.CreateBranch(&config.Branch{Name: branchName, Remote: git.DefaultRemoteName, Merge: plumbing.Master}); err != nil {
-			return err
+			return fmt.Errorf("failed to create branch %q: %w", branchName, err)
 		}
 		return nil
-	default:
-		return err
 	case nil:
 		// fallthrough ...
+	default:
+		return fmt.Errorf("failed to clone %q: %w", a.Repo, err)
 	}
-	if inputArtifact.Git.Fetch != nil {
-		refSpecs := make([]config.RefSpec, len(inputArtifact.Git.Fetch))
-		for i, spec := range inputArtifact.Git.Fetch {
+	if len(a.Fetch) > 0 {
+		refSpecs := make([]config.RefSpec, len(a.Fetch))
+		for i, spec := range a.Fetch {
 			refSpecs[i] = config.RefSpec(spec)
 		}
-		fetchOptions := git.FetchOptions{
-			Auth:     auth,
-			RefSpecs: refSpecs,
-			Depth:    inputArtifact.Git.GetDepth(),
+		opts := &git.FetchOptions{Auth: auth, RefSpecs: refSpecs, Depth: depth}
+		if err := opts.Validate(); err != nil {
+			return fmt.Errorf("failed to validate fetch %v: %w", refSpecs, err)
+		}
+		if err = r.Fetch(opts); isFetchErr(err) {
+			return fmt.Errorf("failed to fetch %v: %w", refSpecs, err)
+		}
+	}
+	w, err := r.Worktree()
+	if err != nil {
+		return fmt.Errorf("failed to get work tree: %w", err)
+	}
+	if a.Revision != "" {
+		if err := r.Fetch(&git.FetchOptions{RefSpecs: []config.RefSpec{"refs/heads/*:refs/heads/*"}}); isFetchErr(err) {
```

Comment:
Should this be `refs/heads/*:refs/remotes/origin/*`?

Reply:
Hmm. Passes the tests. Maybe?